default/AI/ShipDesignAI.py

```diff
@@ -1548,6 +1630,102 @@ def _calc_rating_for_name(self):
         return self.structure*self._total_dmg()*(1+self.shields/10)
 
 
+class CarrierShipDesigner(ShipDesigner):  # TODO consider inheriting from MilitaryShipDesigner
+    """Class that implements military designs with fighter parts.
+
+    Extends __init__()
+    Extends _filling_algorithm()
+    Overrides _rating_function()
+    Overrides _calc_rating_for_name()
+    """
+    basename = "Carrier"
+    description = "Carrier"
+    useful_part_classes = WEAPONS | ARMOUR | SHIELDS | FUEL | ENGINES | FIGHTER_HANGAR | FIGHTER_BAY
+    filter_useful_parts = True
+    filter_inefficient_parts = True
+
+    NAMETABLE = "AI_SHIPDESIGN_NAME_MILITARY"
+    NAME_THRESHOLDS = sorted([0, 1000])
+
+    def __init__(self):
+        ShipDesigner.__init__(self)
+        self.additional_specifications.minimum_fuel = 1
+        self.additional_specifications.minimum_speed = 30
+        self.additional_specifications.expected_turns_till_fight = 10 if fo.currentTurn() < 50 else 5
+        self.additional_specifications.minimum_fighter_launchrate = 1
+
+    def _rating_function(self):
+        if self.fighter_capacity < 1:
+            return INVALID_DESIGN_RATING
+        # first, calculate "normal" weapon stuff
+        weapon_dmg = max(self._total_dmg_vs_shields(), self._total_dmg() / 1000)
+        enemy_dmg_max = self.additional_specifications.max_enemy_weapon_strength
+        enemy_dmg_avg = self.additional_specifications.avg_enemy_weapon_strength
+        shield_factor = max(enemy_dmg_max / max(0.01, enemy_dmg_max - self.shields), 1)
+        expected_growth = min(self.additional_specifications.expected_turns_till_fight * self.organic_growth,
```

Comment:
This is copy-paste from `MilitaryShipDesigner`, please solve it by inheritance mentioned in TODO or(and) extracting this code to base class `self._get_expected_growth()`  

same for other copy-pasted blocks